File: test/debug/color_correction.py
import cv2
import numpy as np


class ColorCorrection:
    """Preprocessor which rotates images"""

    sift_features: int
    hws: float
    blur_size: float
    poly_degrees: tuple[int, int, int]

    def __init__(self) -> None:
        self.sift_features = 10000
        self.hws = 30
        self.blur_size = 11
        self.poly_degrees = (7, 1, 1)

    def process(self, img1, img2):
        kpoints1, kpoints2 = self._find_relevant_points(img1, img2)
        kpoints1, kpoints2 = self._non_max_suppression(kpoints1, kpoints2)
        color_factors = self._calculate_color_correction(img1, img2, kpoints1, kpoints2)
        color_factors, Y = self._interpolate_missing_factors(color_factors)
        img2_modified = self._apply_color_correction(img2, color_factors, Y)

        return img1, img2_modified

    def _find_relevant_points(self, img1, img2):
        img1_gray = cv2.cvtColor(img1, cv2.COLOR_BGR2GRAY)
        img2_gray = cv2.cvtColor(img2, cv2.COLOR_BGR2GRAY)

        sift = cv2.SIFT_create(nfeatures=self.sift_features)
        kpts_img1, descs_img1 = sift.detectAndCompute(img1, None)
        kpts_img2, descs_img2 = sift.detectAndCompute(img2, None)
        bfm = cv2.BFMatcher_create(cv2.NORM_L2, crossCheck=True)
        matches = bfm.match(descs_img1, descs_img2)

        # kpts1_img = cv2.drawKeypoints(img1, kpts_img1, None, color=(72, 144, 233))
        # cv2.imshow("f", kpts1_img)
        # cv2.waitKey(0)

        # matches_img = cv2.drawMatches(
        #     img1, kpts_img1, img2, kpts_img2, matches, None, flags=cv2.DrawMatchesFlags_NOT_DRAW_SINGLE_POINTS
        # )
        # cv2.imshow("f", matches_img)
        # cv2.waitKey(0)

        matched1 = np.array([kpts_img1[m.queryIdx] for m in matches])
        matched2 = np.array([kpts_img2[m.trainIdx] for m in matches])
        return matched1, matched2

    def _non_max_suppression(self, points1, points2):
        valid = []

        for i in range(len(points1)):
            point_i = points1[i].pt
            is_max = True
            for j in range(len(valid)):
                point_j = points1[valid[j]].pt
                dist = (point_i[0] - point_j[0]) ** 2 + (point_i[1] - point_j[1]) ** 2
                if dist < self.hws**2:
                    is_max = False
                    break

            if is_max and points1[i].size < 5:
                valid.append(i)

        return points1[valid], points2[valid]

    def _calculate_color_correction(self, img1, img2, features1, features2):
        color_factors = np.zeros(shape=(256, 3), dtype=np.float32)
        factor_counter = np.zeros(shape=(256, 3), dtype=np.float32)

        img1_ycrcb = cv2.cvtColor(img1, cv2.COLOR_BGR2YCrCb)
        img2_ycrcb = cv2.cvtColor(img2, cv2.COLOR_BGR2YCrCb)

        img1_ycrcb[:, :, 0] = img1_ycrcb[:, :, 0] / 255 * (235 - 16) + 16
        img1_ycrcb[:, :, 1] = img1_ycrcb[:, :, 1] / 255 * (240 - 16) + 16
        img1_ycrcb[:, :, 2] = img1_ycrcb[:, :, 2] / 255 * (240 - 16) + 16
        img2_ycrcb[:, :, 0] = img2_ycrcb[:, :, 0] / 255 * (235 - 16) + 16
        img2_ycrcb[:, :, 1] = img2_ycrcb[:, :, 1] / 255 * (240 - 16) + 16
        img2_ycrcb[:, :, 2] = img2_ycrcb[:, :, 2] / 255 * (240 - 16) + 16

        img_1_ycrcb_smooth = cv2.GaussianBlur(img1_ycrcb, (self.blur_size, self.blur_size), sigmaX=0.5, sigmaY=0.5)
        img_2_ycrcb_smooth = cv2.GaussianBlur(img2_ycrcb, (self.blur_size, self.blur_size), sigmaX=0.5, sigmaY=0.5)
        height1, width1 = img_1_ycrcb_smooth.shape[:2]
        height2, width2 = img_2_ycrcb_smooth.shape[:2]

        for f1, f2 in zip(features1, features2):
            p1, p2 = (round(f1.pt[0]), round(f1.pt[1])), (round(f2.pt[0]), round(f2.pt[1]))

            hws_1 = min(p1[0], width1 - p1[0] - 1, p1[1], height1 - p1[1] - 1)
            hws_2 = min(p2[0], width2 - p2[0] - 1, p2[1], height2 - p2[1] - 1)
            curr_hws = min(self.hws, hws_1, hws_2)

            img_1_section = img_1_ycrcb_smooth[
                p1[1] - curr_hws : p1[1] + curr_hws + 1, p1[0] - curr_hws : p1[0] + curr_hws + 1, :
            ]
            img_2_section = img_2_ycrcb_smooth[
                p2[1] - curr_hws : p2[1] + curr_hws + 1, p2[0] - curr_hws : p2[0] + curr_hws + 1, :
            ]

            color_diff = img_1_section / img_2_section

            for row in range(2 * curr_hws + 1):
                for col in range(2 * curr_hws + 1):
                    idx = min(234, max(15, round(img_2_section[row, col, 0])))
                    color_factors[idx, 0] = color_factors[idx, 0] + color_diff[row, col, 0]
                    factor_counter[idx, 0] = factor_counter[idx, 0] + 1

                    for chan in [1, 2]:
                        idx = min(239, max(15, round(img_2_section[row, col, chan])))
                        color_factors[idx, chan] = color_factors[idx, chan] + color_diff[row, col, chan]
                        factor_counter[idx, chan] = factor_counter[idx, chan] + 1

        color_factors /= factor_counter
        return color_factors

    def _interpolate_missing_factors(self, color_factors):
        color_factors = np.nan_to_num(color_factors)
        _, Y = np.meshgrid(range(3), range(256))

        p_y = np.polyfit(Y[color_factors[:, 0] > 0, 0], color_factors[color_factors[:, 0] > 0, 0], self.poly_degrees[0])
        p_cr = np.polyfit(
            Y[color_factors[:, 1] > 0, 0], color_factors[color_factors[:, 1] > 0, 1], self.poly_degrees[1]
        )
        p_cb = np.polyfit(
            Y[color_factors[:, 2] > 0, 0], color_factors[color_factors[:, 2] > 0, 2], self.poly_degrees[2]
        )

        range_y = list(range(16, 236))
        range_cr_cb = list(range(16, 241))
        color_factors[range_y, 0] = np.polyval(p_y, range_y)
        color_factors[range_cr_cb, 1] = np.polyval(p_cr, range_cr_cb)
        color_factors[range_cr_cb, 2] = np.polyval(p_cb, range_cr_cb)

        return color_factors, Y

    def _apply_color_correction(self, img2, color_factors, Y):
        color_maps = np.clip(np.around(color_factors[:, :] * Y), 0, 255)

        img2_ycrcb = cv2.cvtColor(img2, cv2.COLOR_BGR2YCrCb)
        img2_ycrcb[:, :, 0] = img2_ycrcb[:, :, 0] / 255 * (235 - 16) + 16
        img2_ycrcb[:, :, 1] = img2_ycrcb[:, :, 1] / 255 * (240 - 16) + 16
        img2_ycrcb[:, :, 2] = img2_ycrcb[:, :, 2] / 255 * (240 - 16) + 16

        img_2_new = np.zeros_like(img2_ycrcb)
        img_2_new[:, :, 0] = color_maps[img2_ycrcb[:, :, 0], 0]
        img_2_new[:, :, 1] = color_maps[img2_ycrcb[:, :, 1], 1]
        img_2_new[:, :, 2] = color_maps[img2_ycrcb[:, :, 2], 2]

        img_2_new[:, :, 0] = (img_2_new[:, :, 0] - 16) / (235 - 16) * 255
        img_2_new[:, :, 1] = (img_2_new[:, :, 1] - 16) / (240 - 16) * 255
        img_2_new[:, :, 2] = (img_2_new[:, :, 2] - 16) / (240 - 16) * 255

        img_2_new_bgr = cv2.cvtColor(img_2_new.astype(np.uint8), cv2.COLOR_YCrCb2BGR)
        return img_2_new_bgr


img1 = cv2.imread(r"D:\UserFolders\Downloads\Translated.jpg")
img2 = cv2.imread(r"D:\UserFolders\Downloads\Warped_Overtune.jpg")

cc = ColorCorrection()
# img1_cc, img2_cc = cc.process(img1, img2)

color_maps = np.loadtxt("test/debug/color_maps.txt", delimiter=",", dtype=np.uint8)
color_maps -= 1
img2_ycrcb = cv2.cvtColor(img2, cv2.COLOR_BGR2YCrCb)
img2_ycrcb[:, :, 0] = img2_ycrcb[:, :, 0] / 255 * (235 - 16) + 16
img2_ycrcb[:, :, 1] = img2_ycrcb[:, :, 1] / 255 * (240 - 16) + 16
img2_ycrcb[:, :, 2] = img2_ycrcb[:, :, 2] / 255 * (240 - 16) + 16

img_2_new = np.zeros_like(img2_ycrcb)
img_2_new[:, :, 0] = color_maps[img2_ycrcb[:, :, 0], 0]
img_2_new[:, :, 1] = color_maps[img2_ycrcb[:, :, 1], 2]
img_2_new[:, :, 2] = color_maps[img2_ycrcb[:, :, 2], 1]

img_2_new[:, :, 0] = (img_2_new[:, :, 0] - 16) / (235 - 16) * 255
img_2_new[:, :, 1] = (img_2_new[:, :, 1] - 16) / (240 - 16) * 255
img_2_new[:, :, 2] = (img_2_new[:, :, 2] - 16) / (240 - 16) * 255

img2_cc = cv2.cvtColor(img_2_new.astype(np.uint8), cv2.COLOR_YCrCb2BGR)


cv2.imshow("t", img2_cc)
cv2.waitKey(0)

# A = np.array([2, 4, 6, 8, 3, 5, 7, 9]).reshape(2, 4)
# B = np.ones_like(A) * 10
# B[0, 0] = 0
# print(A / B)

File: test/debug/color_correction.m
clear;

%% parameters

% half window size - determines the minimum distance between used features
% points as well as the size of the window around a feature point which is
% used for the color adjustment calculations
hws = 30; 

% image blur size used to reduce the effect of noise
blur_size = 10;

% degrees of polynomials which are fitted to the measured coefficients for
% the Y, Cb and Cr channels
poly_degree_y = 7;
poly_degree_cb_cr = 1;


%% prepare data

% load images
img_1 = imread('Translated.jpg');
img_2 = imread('Warped_Overtune.jpg');

% convert to grayscale
img_1_gray = rgb2gray(img_1);
img_2_gray = rgb2gray(img_2); 


%% feature extraction and matching

% get SURF features
features_1 = detectSURFFeatures(img_1_gray);
features_2 = detectSURFFeatures(img_2_gray);

% extract features/points
[f_1,points_1] = extractFeatures(img_1_gray,features_1);
[f_2,points_2] = extractFeatures(img_2_gray,features_2);

% match features
indexPairs = matchFeatures(f_1,f_2);
matched_1 = points_1(indexPairs(:,1),:);
matched_2 = points_2(indexPairs(:,2),:);

% filter matches according to point distances
valid = [];
for i = 1:size(indexPairs,1)
	% non-max suppression: check if there already is a valid feature
	% within a hws px radius
	is_max = true;
	for j = 1:size(valid,2)
	    diff = matched_1(i).Location - matched_1(valid(j)).Location;
	    dist = diff * diff';
	    if dist < hws^2
            is_max = false;
            break;
	    end
	end
	if is_max && matched_1(i).Scale < 5
	    % save new feature index
	    valid = [valid, i];
	end
end
matched_1 = matched_1(valid);
matched_2 = matched_2(valid);


%% calculate color correction transform

color_factors = double(zeros(256,3));
factor_counter = double(zeros(256,3));

% convert to ycbcr space
img_1_ycbcr = rgb2ycbcr(img_1);
img_2_ycbcr = rgb2ycbcr(img_2);

% smooth images to reduce noise effect
img_1_ycbcr_smooth = imgaussfilt(img_1_ycbcr,blur_size);
img_2_ycbcr_smooth = imgaussfilt(img_2_ycbcr,blur_size);

for i = 1:size(valid,2)
    % collect samples around feature points
    p_1 = round(matched_1(i).Location);
    p_2 = round(matched_2(i).Location);

    hws_1 = min(min(p_1(1)-1, size(img_1_ycbcr_smooth,1) - p_1(1)), min(p_1(2)-1, size(img_1_ycbcr_smooth, 2) - p_1(2)));
    hws_2 = min(min(p_2(1)-1, size(img_2_ycbcr_smooth,1) - p_2(1)), min(p_2(2)-1, size(img_2_ycbcr_smooth, 2) - p_2(2)));
    curr_hws = min(hws, min(hws_1, hws_2));

    img_1_section = img_1_ycbcr_smooth(p_1(2)-curr_hws:p_1(2)+curr_hws, p_1(1)-curr_hws:p_1(1)+curr_hws, :);
    img_2_section = img_2_ycbcr_smooth(p_2(2)-curr_hws:p_2(2)+curr_hws, p_2(1)-curr_hws:p_2(1)+curr_hws, :);

    % calculate the color difference
    color_diff = (double(img_1_section) ./ double(img_2_section));
    
    % save the factors for the respective y, cb and cr channels
    for row = 1:2*curr_hws+1
        for col = 1:2*curr_hws+1
            idx = min(235,max(16,round(img_2_section(row,col,1))));
                color_factors(idx, 1) = ...
                    color_factors(idx, 1) + color_diff(row,col,1);
                factor_counter(idx, 1) = ...
                    factor_counter(idx, 1) + 1;
            for chan = 2:3
                idx = min(240,max(16,round(img_2_section(row,col,chan))));
                color_factors(idx, chan) = ...
                    color_factors(idx, chan) + color_diff(row,col,chan);
                factor_counter(idx, chan) = ...
                    factor_counter(idx, chan) + 1;
            end
        end
    end
end
color_factors = color_factors ./ factor_counter;

% some bins might be empty -> generate missing factors via polynomial
% interpolation
color_factors(isnan(color_factors)) = 0;

[~,Y]=meshgrid(1:3,1:256);
p_y = polyfit(Y(color_factors(:,1)>0,1),color_factors(color_factors(:,1)>0,1), poly_degree_y);
p_cb = polyfit(Y(color_factors(:,2)>0,1),color_factors(color_factors(:,2)>0,2), poly_degree_cb_cr);
p_cr = polyfit(Y(color_factors(:,3)>0,1),color_factors(color_factors(:,3)>0,3), poly_degree_cb_cr);

figure(1); subplot(3,1,1); plot(linspace(1,256,256),color_factors(:,1)); hold on;
subplot(3,1,2); plot(linspace(1,256,256),color_factors(:,2)); hold on;
subplot(3,1,3); plot(linspace(1,256,256),color_factors(:,3)); hold on;

range_y = 16:1:235;
color_factors(range_y,1) = polyval(p_y, range_y);
range_cb_cr = 16:1:240;
color_factors(range_cb_cr,2) = polyval(p_cb, range_cb_cr);
color_factors(range_cb_cr,3) = polyval(p_cr, range_cb_cr);

color_factors(color_factors < 0) = 0;
color_factors(color_factors > 10) = 10;

figure(1); subplot(3,1,1); plot(linspace(1,256,256),color_factors(:,1));
subplot(3,1,2); plot(linspace(1,256,256),color_factors(:,2));
subplot(3,1,3); plot(linspace(1,256,256),color_factors(:,3));


%% apply color correction

% create a ycbcr color map to apply to img_2
color_maps = min(256,max(1,round(color_factors(:,:) .* Y)));

% apply color map
img_2_new(:,:,1) = reshape(color_maps(img_2_ycbcr(:,:,1),1),size(img_2_ycbcr(:,:,1)));
img_2_new(:,:,2) = reshape(color_maps(img_2_ycbcr(:,:,2),2),size(img_2_ycbcr(:,:,2)));
img_2_new(:,:,3) = reshape(color_maps(img_2_ycbcr(:,:,3),3),size(img_2_ycbcr(:,:,3)));

% convert to rgb
img_2_new = ycbcr2rgb(uint8(img_2_new));


%% output

% plot features onto images
set(groot,'defaultLineLineWidth',2.0)
figure(2);
imshow(img_1); hold on;
plot(matched_1);
img_1 = getframe;

imshow(img_2); hold on;
plot(matched_2);
img_2 = getframe;

imshow(img_2_new); hold on;
plot(matched_2);
img_2_new = getframe;
close;

figure(2);
montage({img_2.cdata,img_1.cdata,img_2_new.cdata,img_1.cdata});
